proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,1051 @@
+
+The proposal up to this point has introduced the visible non-dependent case.
+The visible dependent quantifier is analogous ::
+
+  type FVis :: forall k -> @U k -> @U Type
+  type family FVis k (a :: k) :: Type
+
+
+  type DVis :: forall k -> @M k -> @M Type
+  data DVis k (a :: k) :: Type
+
+Now consider the invisible dependent version of the above two types ::
+
+  type FInvis :: forall k. @U k -> @U Type
+  type family FInvis (a :: k) :: Type
+
+  type DInvis :: forall k. @M k -> @M Type
+  data DInvis (a :: k) :: Type
+
+Notice that the forall itself is annotated in both cases. The treatment of
+invisible quantifiers is necessary to properly handle higher-rank programs. To
+illustrate why, consider the following program ::
+
+  type D :: forall (f :: forall k. @U k -> @U Type) -> @M Type
+  data D f = D (f Bool) (f 0)
+
+  type F :: forall k. @U k -> @U Type
+  type family F a where
+    F 0 = Int
+    F Bool = Char
+
+  p :: D F
+  p = D 'c' 0
+
+Here, ``D`` has a rank-2 kind and its argument is a function. To be able to
+pass in ``F``, the forall must be unmatchable in ``D``'s argument.
+
+A matchability variable introduced in a ``forall`` telescope is in scope in the
+body of the ``forall``, but not in its matchability. That is, ``forall m x. @m x``
+is invalid, but ``forall m. @U forall x. @m x`` is valid. The same applies to the
+visible case: ```forall m -> @m Type``` is rejected.
+
+.. _Inference:
+
+Meaning of an annotation-free ``->``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Even though this proposal introduces a way to annotate arrows, in many cases the
+annotations can be inferred. The primary aim of inference is to ease the
+transition as most programs written today can be unambiguously inferred.
+
+The meaning of ``(->)`` depends on the context in which it is written. Below is
+a list of the different contexts with examples. Some of these are a result of a
+*defaulting* mechanism, and some are *hard rules*. The first category can be
+overridden with an appropriate annotation, but the second can not. The
+defaulting mechanism first invents a unification variable for the matchability,
+and only applies the defaulting rule here when the variable remains
+unconstrained after constraint solving.
+
+Data types
+##########
+
+The kind arrows of data types (and data families) are all matchable.
+
+::
+
+  -- inferred:  Type -> @M Type
+  type Maybe :: Type -> Type
+  data Maybe a = ...
+
+here, users are not required to specify ``Type -> @M Type``, as this information
+can be inferred from the data declaration itself. This is a *hard rule*, and
+it is an error to write ::
+
+  type Maybe :: Type -> @U Type
+  data Maybe a = ...
+
+Data constructors
+#################
+
+Data constructors are matchable. This means that using either syntax ::
+
+  data Maybe a = Nothing | Just a
+
+  data Maybe a where
+    Nothing :: Maybe a
+    Just :: a -> Maybe a
+
+``Just :: a -> @M Maybe a``. GHC already eta-expands data constructors
```

Comment:
Let me answer your questions in order:

- Yes, this is in part preparation for a dependent future, but also it makes promotion internally more consistent, because this way both `Just` the data constructor and `Just` the type constructor are matchable. But thanks to eta-expansion, we don't need to worry about this in terms (and in the future we can think about eta-expansion in types too, which would improve ergonomics, but currently type lambdas don't exist).
- No, everything is unmatchable (I think my answer to the previous question covered this one too)
- As far as I'm aware, GHC already does eta expansion during typechecking. This was introduced in the LinearTypes implementation (the [wiki](https://gitlab.haskell.org/ghc/ghc/-/wikis/linear-types/implementation#data-constructors-are-polymorphic) describes the details). Notably, the expansion happens whether or not LinearTypes is enabled, so there is no additional implementation burden here.

Reply:
> Yes, this is in part preparation for a dependent future, but also it makes promotion internally more consistent, because this way both Just the data constructor and Just the type constructor are matchable.

The preparation for dependent haskell I understand, though I think it would be helpful to point that out in the text. While it makes sense to me to talk about the matchability of `Just`-the-datacon so we understand how it behaves when promoted, this section also says that `map`-the-function expects unmatchable arguments. Functions are not promoted, so why does matchability enter the picture for `map`? It sounds like the answer is that we expect it to be relevant eventually, so we might as well sort it out now. That's a perfectly good answer, but it would help to make that explicit, otherwise I'm left feeling like I may be missing a very important detail.

> As far as I'm aware, GHC already does eta expansion during typechecking. This was introduced in the LinearTypes implementation (the wiki describes the details). Notably, the expansion happens whether or not LinearTypes is enabled, so there is no additional implementation burden here.

Ah yes, thanks for the reminder, I had forgotten about this detail of LinearTypes.

Reply:
I updated the text to be clearer about this point!